ingestion/src/kldata.rs

```diff
@@ -0,0 +1,279 @@
+use crate::{
+    permissions::{timeseries_is_open, PermitTable},
+    Datum, Error, PooledPgConn,
+};
+use chrono::{DateTime, Utc};
+use regex::Regex;
+use std::{
+    collections::HashMap,
+    fmt::Debug,
+    str::{FromStr, Lines},
+    sync::{Arc, RwLock},
+};
+
+pub struct ObsinnObs {
+    timestamp: DateTime<Utc>,
+    id: ObsinnId,
+    value: f32,
+}
+
+pub struct ObsinnChunk {
+    observations: Vec<ObsinnObs>,
+    station_id: i32, // TODO: change name here to nationalnummer?
+    type_id: i32,
+}
+
+#[derive(Debug, Clone)]
+struct ObsinnId {
+    param_code: String,
+    sensor_and_level: Option<(i32, i32)>,
+}
+
+fn parse_meta_field<T: FromStr>(field: &str, expected_key: &'static str) -> Result<T, Error>
+where
+    <T as FromStr>::Err: Debug,
+{
+    let (key, value) = field.split_once('=').ok_or_else(|| {
+        Error::Parse(format!(
+            "unexpected field in kldata header format: {}",
+            field
+        ))
+    })?;
+
+    // TODO: Replace assert with error?
+    assert_eq!(key, expected_key);
+
+    let res = value
+        .parse::<T>()
+        .map_err(|_| Error::Parse(format!("invalid number in kldata header for key {}", key)))?;
+
+    Ok(res)
+}
+
+fn parse_meta(meta: &str) -> Result<(i32, i32, usize), Error> {
+    let mut parts = meta.splitn(4, '/');
+
+    let next_err = || Error::Parse("kldata header terminated early".to_string());
+
+    if parts.next().ok_or_else(next_err)? != "kldata" {
+        return Err(Error::Parse(
+            "kldata indicator missing or out of order".to_string(),
+        ));
+    }
+
+    let station_id = parse_meta_field(parts.next().ok_or_else(next_err)?, "nationalnr")?;
+    let type_id = parse_meta_field(parts.next().ok_or_else(next_err)?, "type")?;
+    let message_id = parse_meta_field(parts.next().ok_or_else(next_err)?, "messageid")?;
+
+    Ok((station_id, type_id, message_id))
+}
+
+fn parse_columns(cols_raw: &str) -> Result<Vec<ObsinnId>, Error> {
+    // this regex is taken from kvkafka's kldata parser
+    // let col_regex = Regex::new(r"([^(),]+)(\([0-9]+,[0-9]+\))?").unwrap();
+    // it is modified below to capture sensor and level separately, while keeping
+    // the block collectively optional
+    //
+    // TODO: is it possible to reuse this regex even more?
```

Comment:
What exactly do you mean by reuse it even more?

Reply:
This line compiles the regex, which is then reused for the rest of the function. In theory I think it should be possible to compile the regex only once at startup, but it's a pretty minor optimisation